### PR TITLE
derived observables: firsts steps

### PR DESCRIPTION
**Pull request by michi42**
This is a first attempt to implement basic derivation operations on observable values: "map" and "filter".

To not bloat the interfaces, they are implemented for ObservableValue for the moment.

I terms of weak/strong references, the idea is as following:
- subscribe weakly to the upstream observable
- keep a static strong reference as long as at least 1 subscriber is registered
- if no subscriber is registered AND no external reference exists, allow GC

Errors in transformations are caught internally and lead to an update being dropped (do we need error subscriptions?)

### CODE REVIEW

**Comment by kaifox** [comment]
Ok, I was skeptical about the weak ref... but I agree, as long as you keep a strong ref as long there is one subscriber, it should be ok, right?

**Comment by kaifox** [comment]
cf the error subscriptions: Not sure .. .we should probably be consisten with the other interfaces...

**Comment by michi42** [comment]
> as long as you keep a strong ref as long there is one subscriber, it should be ok, right?
That was the idea. It should allow something like
```java
observable.map(i -> i+42).subscribe(object::listen)
```
while not creating a memory leak and being able to GC when something like
```java
observable.map(i -> i+42);
observable.map(i -> i + 2);
observable.map(i -> i);
```
leaves the scope.

**Comment by michi42** [comment]
>  Not sure .. .we should probably be consisten with the other interfaces...
Sure, but should we in general have something like
```java
interface Observable<T> {
   Subscription subscribe(Consumer<T>, Options ...)
   Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
}
```
(with the explicit premise that a delivered exception does NOT kill any subscription or collapse anything)

**Comment by kaifox** [comment]
HI @michi42, thanks for implementing this. I looked through the code and did not find any conceptual things I would not agree.

I think what we should do soon is to document these things really well (also all the edge cases). I hope I can find some time to help you in this ... For such a core library it is definitely important ;-)

**Comment by kaifox** [comment]
> What I meant - should we in general have something like
> 
> ```java
> interface Observable<T> {
>    Subscription subscribe(Consumer<T>, Options ...)
>    Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
> }
Mhmm ... if we do it, I would tend to prefer the `subscribeException` version over the `Observable<Maybe<T>>`...

Another option could be (without thinking too much if it would really work)

```
interface Observable<T> {
   Subscription subscribe(Consumer<T>, Options ...);
   Observable<Throwable> exceptionObservable();
   Observable<Maybe<T>> maybeObservable();
}

... and I definitely would avoid collapsing things ;-) (Except if we introduce another option to do so ;-)

```

**Comment by michi42** [comment]
> Observable<Throwable> exceptions();
But then you could have exceptions of exceptions of exceptions ;-)
but I like the asMaybe()

**Comment by kaifox** [comment]
> > Observable exceptions();
> > But then you could have exceptions of exceptions of exceptions ;-)
> > but I like the asMaybe()

I agree ... I thought the same ...

Could another interface E.g. `ErrorDeflectingObservable` help which would have the two additional methods?

File: src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java [hunk 1]
```diff
@@ -43,13 +45,31 @@ protected DispatchingObservable() {
...
+        listeners.put(listener, subscriptionOptions);
+        if (listenerCount.getAndIncrement() == 0) {
+            firstListenerAdded();
+        }
+    }
+
+    protected void firstListenerAdded() { /* no op */ }
+
+    private void removeListener(Consumer<? super T> listener) {
+        listeners.remove(listener);
```
Comment (andreacalia) [inline src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java]: Should we check if the listener was effectively removed before counting down?

**Comment by andreacalia** [comment]
I also agree with the design :)

File: src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java (hunk 1 above)
Comment (michi42) [reply src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java]: Well, I would say this is an invalid situation - so I'd throw IllegalStateException if the listener was not actually removed.

Comment (kaifox) [reply src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java]: Is ok ... not so sure though ... I think one also could ignore it...(and just not decrement).... (There could be situations where one just would like to make sure that the listeners is not there and one would 'proactively' remove them .... as one has no way to check if its there ...

Comment (michi42) [reply src/main/java/io/github/ossgang/commons/observable/DispatchingObservable.java]: Ok, that's a valid point - someone could keep a Subscription object - if we throw the user must make sure to only call unsubscribe once (which can be a bit tedious in a multithreaded environment), if we are lenient one can just try to call unsubscribe ...

**Pull request merged**